src/storage/src/dedup_pk_cell_based_row_serializer.rs

```diff
@@ -0,0 +1,123 @@
+        column_ids: &[ColumnId],
+    ) -> Self {
+        let pk_indices = pk_indices.iter().cloned().collect::<HashSet<_>>();
+        let dedup_datum_indices = (0..column_descs.len())
+            .filter(|i| {
+                !pk_indices.contains(i) || !column_descs[*i].data_type.mem_cmp_eq_value_enc()
+            })
+            .collect();
+        let dedupped_column_ids = Self::remove_dup_pk_column_ids(&dedup_datum_indices, column_ids);
+        let inner = CellBasedRowSerializer::new(dedupped_column_ids);
+        Self {
+            dedup_datum_indices,
+            inner,
+        }
+    }
+
+    fn filter_by_dedup_datum_indices<'b, I>(
+        dedup_datum_indices: &'b HashSet<usize>,
+        iter: impl Iterator<Item = I> + 'b,
+    ) -> impl Iterator<Item = I> + 'b {
+        iter.enumerate()
+            .filter(|(i, _)| dedup_datum_indices.contains(i))
+            .map(|(_, d)| d)
+    }
+
+    fn remove_dup_pk_datums_by_ref(&self, row: &Row) -> Row {
+        Row(
+            Self::filter_by_dedup_datum_indices(&self.dedup_datum_indices, row.0.iter())
+                .cloned()
+                .collect(),
+        )
+    }
+
+    fn remove_dup_pk_datums(&self, row: Row) -> Row {
+        Row(
+            Self::filter_by_dedup_datum_indices(&self.dedup_datum_indices, row.0.into_iter())
+                .collect(),
+        )
+    }
+
+    fn remove_dup_pk_column_ids(
+        dedup_datum_indices: &HashSet<usize>,
+        column_ids: &[ColumnId],
+    ) -> Vec<ColumnId> {
+        Self::filter_by_dedup_datum_indices(dedup_datum_indices, column_ids.iter())
+            .cloned()
+            .collect()
+    }
+}
+
+impl CellSerializer for DedupPkCellBasedRowSerializer {
+    /// Serialize key and value.
+    fn serialize(&mut self, pk: &[u8], row: Row) -> Result<Vec<(KeyBytes, ValueBytes)>> {
+        let row = self.remove_dup_pk_datums(row);
+        self.inner.serialize(pk, row)
+    }
+
+    /// Serialize key and value. Each column id will occupy a position in Vec. For `column_ids` that
+    /// doesn't correspond to a cell, the position will be None. Aparts from user-specified
+    /// `column_ids`, there will also be a `SENTINEL_CELL_ID` at the end.
+    fn serialize_without_filter(
+        &mut self,
+        pk: &[u8],
+        row: Row,
+    ) -> Result<Vec<Option<(KeyBytes, ValueBytes)>>> {
+        let row = self.remove_dup_pk_datums(row);
+        self.inner.serialize_without_filter(pk, row)
+    }
+
+    /// Different from [`DedupPkCellBasedRowSerializer::serialize`], only serialize key into cell
+    /// key (With column id appended).
+    fn serialize_cell_key(&mut self, pk: &[u8], row: &Row) -> Result<Vec<KeyBytes>> {
+        let row = self.remove_dup_pk_datums_by_ref(row);
+        self.inner.serialize_cell_key(pk, &row)
+    }
+
+    /// Get column ids used by cell serializer to serialize.
+    /// TODO: This should probably not be exposed to user.
+    fn column_ids(&self) -> &[ColumnId] {
```

Comment:
Currently, `CellBasedTable` fetches `column_ids` from cell based row serializer: https://github.com/singularity-data/risingwave/blob/5f769d63be84c35ac745d2cecbf19a32f48d0bd2/src/storage/src/table/cell_based_table.rs#L157-L159
Hence this is required for compatibility.

However, I think it doesn't fit the interface of a serializer, perhaps we should either:
- Store a reference to it in `CellBasedTable`, using `Arc` etc...
- Or, since it is currently only used to serialize pk with column ids, we can create a new serializer which does that, and pass this in to avoid polluting the serializer interface.